src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp

```diff
@@ -300,7 +300,7 @@ void ShenandoahBarrierSetAssembler::load_reference_barrier(MacroAssembler* masm,
     assert(!is_narrow, "phantom access cannot be narrow");
     target = CAST_FROM_FN_PTR(address, ShenandoahRuntime::load_reference_barrier_weak);
   }
-  __ call(target);
+  __ rt_call(target);
```

Comment:
Question: does it make sense to replace `call` with `rt_call` when we are invoking the VM code (C++ code)? Here is what I see the difference of the two: `rt_call` emits code (`auipc` or `movptr`) depending on whether the destination could be found in code cache, while `call` depends on `is_32bit_offset_from_codeache`. So it's still possible for `call` to emit the short `auipc` code if not far even when the target is not there in the code cache like this case. But `rt_call` will always emit a long `movptr` sequence for this case.

Reply:
I found some other places which uses plain calls to leaf, instead of cll_VM_leaf.
It seems like it's a guess that registers don't need push/pop, I don't think such speculation is good.
If it must be there we should have a argument to call_VM_leaf saying we don't want to push/pop.

x86 is this case uses the correct:
` __ call_VM_leaf(CAST_FROM_FN_PTR(address, ShenandoahRuntime::load_reference_barrier_weak), c_rarg0, c_rarg1);`

rt_call for non-code-cache do not need relocation, i.e. movptr.
So rt_call and call is not simply interchangeable, i.e. you now need relocation. 

Yes, we can probably do better here, but as the change is mv/li + jalr to movptr + jalr, there is no regression.
So improvement should be done outside of this PR.

Reply:
Yes!